### PR TITLE
Add QuantizedDBuffer for MXFP8 plane storage

### PR DESCRIPTION
**Pull request by wujingyue**
Adds `QuantizedDBuffer` to manage MXFP8 data and scale planes for MFSDP. Scale ownership follows the data layout, with TE padding added when constructing compute wrappers. `get_tensor_view()` returns compact, unswizzled views for direct in-place quantization; `get_local_tensor()` pads scales for TE GEMM preparation. Supports storage lifecycle management and redistribution across ranks.

Tests compare multiple tensors against independent TE quantization, verify scale shapes and storage aliases, compare rowwise and columnwise GEMMs after all-gather, and exercise redistribution through a preallocated destination and back to replicated storage. The test module skips when TE or MXFP8 support is unavailable.

MFSDP parameter integration and end-to-end training tests are in #7114, stacked on this PR.

Validation: Black, isort, pylint, Ruff, and whitespace checks passed. The revised buffer tests passed on four GB200 ranks (5 tests per rank), including compact quantization, GEMM parity after all-gather, redistribution, view aliasing, and direct all-gather with and without a preallocated destination.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+"""Transformer Engine MXFP8 distributed buffers composed from physical planes."""
+
+from collections.abc import Iterable
+
+import torch
+import transformer_engine_torch as tex
+from torch.distributed import DeviceMesh
+from torch.distributed.tensor.placement_types import Placement
+from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer, MXFP8Tensor
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: MCore currently treats Transformer Engine as an optional dependency. Could we adopt a similar approach to allow this to work when TE is not installed?

[https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/extensions/transformer_engine.py#L71](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/extensions/transformer_engine.py#L71)

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+        """Quantize the local FP32 master shard into MXFP8 grouped planes."""
+        for index in range(len(self.rowwise_data.layout.tensor_shapes)):
+            tensor = self.get_local_tensor(index)
+            rowwise_scale = self.rowwise_scale.get_local_tensor(index)
+            columnwise_scale = self.columnwise_scale.get_local_tensor(index)
+            tensor.quantize_(main_weight.get_local_tensor(index))
+            assert tensor._rowwise_scale_inv is not None
+            assert tensor._columnwise_scale_inv is not None
+            # Unpadded wrapper scales already alias their destination planes.
+            if tensor._rowwise_scale_inv.shape != rowwise_scale.shape:
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: When `tensor._rowwise_scale_inv.shape` matches `rowwise_scale.shape`, shouldn’t we initialize `rowwise_scale` from `tensor._rowwise_scale_inv`?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+            columnwise_data=self.columnwise_data.get_local_tensor(index),
+            columnwise_scale_inv=_pad_columnwise_scale(columnwise_scale),
+            fp8_dtype=_MXFP8_DTYPE,
+            quantizer=_MXFP8_QUANTIZER,
+            with_gemm_swizzled_scales=False,
+            device=rowwise_data.device,
+        )
+
+    def quantize_(self, main_weight: DBuffer) -> None:
+        """Quantize the local FP32 master shard into MXFP8 grouped planes."""
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: Should we consider rejecting quantization for placements other than `BlockAtomic` and `Replicate`?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py (hunk 2 above)
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: Curious: why don’t we need to copy `tensor._rowwise_scale_inv` into `rowwise_scale` when their shapes match?

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    source = _grouped(mesh, [Replicate()], distributed_setup.device)
+    for plane in source.planes:
+        plane.local_buffer.copy_(torch.arange(plane.local_buffer.numel(), device=plane.device))
+    destination = _grouped(mesh, [Flat()], distributed_setup.device)
+    result = source.redistribute([Flat()], out=destination)
+
+    assert result is destination
+    for result_plane, source_plane in zip(result.planes, source.planes):
+        torch.testing.assert_close(
+            result_plane.allgather(0).local_buffer, source_plane.local_buffer
```
Comment (shjwudp) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Looks great! Should we also add unit tests for quantization to verify that `QuantizedDBuffer.quantize_` is numerically equivalent to quantizing an `MXFP8Tensor` independently?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: I'll try to put a guard on the user of this module. quantized_dbuffer itself has no reason to exist without te.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py (hunk 3 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: Added a different check. I'll refine the check when we also support Flat -- refining it now is a bit speculative for me.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py (hunk 2 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: Added a code comment.

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    [Replicate(), BlockAtomic(32), Flat(), Shard(0)],
+    ids=["replicated", "block_atomic", "flat", "shard"],
+)
+@pytest.mark.parametrize("shape", [(64, 64), (128, 128)], ids=["padded_scales", "unpadded_scales"])
+def test_quantized_dbuffer_quantization_matches_te(distributed_setup, placement, shape):
+    """Quantized data and compact scales match independently quantized TE tensors."""
+    import transformer_engine_torch as tex
+    from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer
+
+    if torch.cuda.get_device_capability(distributed_setup.device)[0] < 10:
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Easier way for this?

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+).QuantizedDBuffer
+
+
+def _grouped(mesh, placements, device: torch.device) -> QuantizedDBuffer:
+    return QuantizedDBuffer(mesh, placements, [(64, 64)], device)
+
+
+@pytest.mark.parametrize(
+    "placement",
+    [Replicate(), BlockAtomic(32), Flat(), Shard(0)],
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Too much. Just hard-code to BlockAtomic(32) for now.

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    """Quantized data and compact scales match independently quantized TE tensors."""
+    import transformer_engine_torch as tex
+    from transformer_engine.pytorch.tensor.mxfp8_tensor import MXFP8Quantizer
+
+    if torch.cuda.get_device_capability(distributed_setup.device)[0] < 10:
+        pytest.skip("MXFP8 quantization coverage requires Blackwell-or-newer GPUs.")
+    if isinstance(placement, Shard):
+        shape = (shape[0] * distributed_setup.world_size, shape[1])
+    mesh = init_device_mesh(distributed_setup.device.type, (distributed_setup.world_size,))
+    grouped = QuantizedDBuffer(mesh, [placement], [shape], distributed_setup.device)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Harden the test to test multiple tensors?

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    Flat,
+)
+
+QuantizedDBuffer = pytest.importorskip(
+    "megatron.core.distributed.fsdp.src.megatron_fsdp.experimental.quantized_dbuffer",
+    reason="Transformer Engine MXFP8 support is required.",
+).QuantizedDBuffer
+
+
+def _grouped(mesh, placements, device: torch.device) -> QuantizedDBuffer:
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Inline it. https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    result = grouped.allgather(0)
+
+    assert result.rowwise_data.placements == (Replicate(),)
+    for plane in result.planes:
+        assert plane.local_buffer.view(mesh.size(), -1)[0].eq(0).all()
+        assert plane.local_buffer.view(mesh.size(), -1)[1].eq(1).all()
+
+
+def test_quantized_dbuffer_redistributes_into_matching_destinations(distributed_setup):
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: May just merge this with the preceding test and verify `redistribute` only

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py (hunk 4 above)
Comment (wujingyue) [reply tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Done

File: tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    result = source.redistribute([Flat()], out=destination)
+
+    assert result is destination
+    for result_plane, source_plane in zip(result.planes, source.planes):
+        torch.testing.assert_close(
+            result_plane.allgather(0).local_buffer, source_plane.local_buffer
+        )
+
+
+def test_quantized_dbuffer_derives_scale_shards_from_data(distributed_setup):
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_quantized_dbuffer.py]: Still needed? Could be merged into test_quantized_dbuffer_quantization_matches_te after we change that test to test multiple shapes in one QuantizedDBuffer.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+_MXFP8_BLOCK_SIZE = 32
+
+
+def effective_dtype(tensor: torch.Tensor) -> torch.dtype:
+    """Return MFSDP's storage dtype for a parameter."""
+    return torch.uint8 if isinstance(tensor, MXFP8Tensor) else tensor.dtype
+
+
+def _rowwise_scale_layout(data_layout: GlobalLayout) -> GlobalLayout:
+    """Derive rowwise scale coordinates from the data layout."""
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: This doesn't say where the input data layout come from. I guess it's the data layout of rowwise_data and columnwise_data?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+def _pad_rowwise_scale(scale: torch.Tensor) -> torch.Tensor:
+    """Pad rowwise scales to TE's physical allocation shape."""
+    shape = _MXFP8_QUANTIZER.get_scale_shape(
+        (scale.shape[0], scale.shape[1] * _MXFP8_BLOCK_SIZE), columnwise=False
+    )
+    if scale.shape == shape:
+        return scale
+    padded = torch.zeros(shape, dtype=scale.dtype, device=scale.device)
+    padded[: scale.shape[0], : scale.shape[1]].copy_(scale)
+    return padded
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/quantized_dbuffer.py]: Isn't there an easier way to pad?

**Review by shjwudp — APPROVED** [review]
LGTM!
